internal/datacoord/sync_segments_scheduler.go

```diff
@@ -0,0 +1,155 @@
+// Licensed to the LF AI & Data foundation under one
+// or more contributor license agreements. See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership. The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License. You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package datacoord
+
+import (
+	"sync"
+	"time"
+
+	"github.com/samber/lo"
+	"go.uber.org/zap"
+
+	"github.com/milvus-io/milvus/internal/proto/datapb"
+	"github.com/milvus-io/milvus/pkg/log"
+	"github.com/milvus-io/milvus/pkg/util/logutil"
+	"github.com/milvus-io/milvus/pkg/util/typeutil"
+)
+
+type SyncSegmentsScheduler struct {
+	quit chan struct{}
+	wg   sync.WaitGroup
+
+	meta           *meta
+	channelManager ChannelManager
+	sessions       SessionManager
+}
+
+func newSyncSegmentsScheduler(m *meta, channelManager ChannelManager, sessions SessionManager) *SyncSegmentsScheduler {
+	return &SyncSegmentsScheduler{
+		quit:           make(chan struct{}),
+		wg:             sync.WaitGroup{},
+		meta:           m,
+		channelManager: channelManager,
+		sessions:       sessions,
+	}
+}
+
+func (sss *SyncSegmentsScheduler) Start() {
+	sss.quit = make(chan struct{})
+	ticker := time.NewTicker(Params.DataCoordCfg.SyncSegmentsInterval.GetAsDuration(time.Second))
+	sss.wg.Add(1)
+
+	go func() {
+		defer logutil.LogPanic()
+		defer sss.wg.Done()
+
+		for {
+			select {
+			case <-sss.quit:
+				log.Info("sync segments scheduler quit")
+				ticker.Stop()
+				return
+			case <-ticker.C:
+				sss.SyncSegmentsForCollections()
+			}
+		}
+	}()
+	log.Info("SyncSegmentsScheduler started...")
+}
+
+func (sss *SyncSegmentsScheduler) Stop() {
+	close(sss.quit)
+	sss.wg.Wait()
+}
+
+func (sss *SyncSegmentsScheduler) SyncSegmentsForCollections() {
+	collIDs := sss.meta.ListCollections()
+	for _, collID := range collIDs {
+		collInfo := sss.meta.GetCollection(collID)
+		if collInfo == nil {
+			log.Warn("collection info is nil, skip it", zap.Int64("collectionID", collID))
+			continue
+		}
+		pkField, err := typeutil.GetPrimaryFieldSchema(collInfo.Schema)
+		if err != nil {
+			log.Warn("get primary field from schema failed", zap.Int64("collectionID", collID),
+				zap.Error(err))
+			continue
+		}
+		for _, channelName := range collInfo.VChannelNames {
```

Comment:
Maybe we can execute concurrently at the channel level?

Reply:
I thinks it's no necessary. I already add the concurrently for load pk in datanode.